### PR TITLE
Jyaou/mongo setup

### PR DESCRIPTION
**Pull request by JyaouShingan**

### CODE REVIEW

File: src/database/mongo-service.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            return conn.db.databaseName == database;
+        });
+        if (existConnection) return existConnection;
+        const connection = await this.client.createConnection(`${this.uri}/${database}`, {useNewUrlParser: true});
+        this.logger.info('Connected to database');
+        return connection;
+    }
+}
+
+export const MongoDB = new MongoDBService();
```
Comment (TeakayAero) [inline src/database/mongo-service.ts]: Do export default for files only have one class

File: src/app.ts [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
 const logger = getLogger("app.js");
 
-if (!config || !config.discordToken) {
-    logger.error(`Missing botconfig.json or "discordToken" in json`);
+async function main() {
+    if (!config || !config.discordToken) {
+        logger.error(`Missing botconfig.json or "discordToken" in json`);
+    }
+
+    if (await MongoDB.start()) {
```
Comment (TeakayAero) [inline src/app.ts:14]: false handle?

File: src/bilibili-api.ts [hunk 3]
```diff
@@ -49,24 +48,23 @@ export function search(keyword: string, limit?: number): Promise<SearchSongEntit
...
 
-        if (!rawSongs) return [];
-        return rawSongs.map((raw) => {
-            return new SearchSongEntity()
-                .setTitle(raw['title'])
-                .setAuthor(raw['author'])
-                .setVideoId(raw['param'])
-                .setPlay(raw['play']);
-        });
+    const response = await request(req);
```
Comment (TeakayAero) [inline src/bilibili-api.ts:52]: future: probably should type this

File: src/commands/load.ts [hunk 4]
```diff
@@ -22,43 +22,34 @@ export class LoadCommand extends BaseCommand {
...
     }
 
     private async load(message: Message, guild: GuildManager, collection?: string) {
-        if (!fs.existsSync('./playlist')) {
-            fs.mkdirSync('./playlist');
-            message.reply('Nothing here yet');
-            return;
+        message.reply('Start loading playlist');
+        const songs = await guild.datasource.loadFromPlaylist(message.author, collection);
+        for (const song of songs) {
```
Comment (TeakayAero) [inline src/commands/load.ts:35]: future: threshold

File: src/commands/load.ts [hunk 5]
```diff
@@ -22,43 +22,34 @@ export class LoadCommand extends BaseCommand {
...
             throw CommandException.UserPresentable('Too many arguments, expected argument: 1');
         }
-        message.reply("Finished loading from the playlist");
     }
 
     helpMessage(): string {
         return 'Usage: load <list-name>';
     }
 
     private async load(message: Message, guild: GuildManager, collection?: string) {
```
Comment (TeakayAero) [inline src/commands/load.ts]: TODO: migration script

File: src/commands/load.ts [hunk 6]
```diff
@@ -22,43 +22,34 @@ export class LoadCommand extends BaseCommand {
...
+        for (const song of songs) {
+            song.streamer.start();
+            guild.playlist.push(song);
         }
+        message.reply("Finished loading from the playlist");
 
-        const playlistName = collection ? `./playlist/${collection}` : './playlist/default';
-        if (!fs.existsSync(playlistName)) {
-            message.reply('The playlist does not exist');
+        if (songs.length === 0) {
```
Comment (TeakayAero) [inline src/commands/load.ts:40]: Any message handle?

File: src/commands/promote.ts [hunk 7]
```diff
@@ -14,11 +14,13 @@ export class PromoteCommand extends BaseCommand {
             throw CommandException.UserPresentable(this.helpMessage());
         }
 
-        const index = parseInt(args.shift());
-        if (isNaN(index) || !Number.isInteger(index)) {
+        const userIndex = parseInt(args.shift());
+        if (isNaN(userIndex) || !Number.isInteger(userIndex)) {
```
Comment (TeakayAero) [inline src/commands/promote.ts]: Future: utilize this

File: src/commands/promote.ts [hunk 8]
```diff
@@ -14,11 +14,13 @@ export class PromoteCommand extends BaseCommand {
...
             throw CommandException.UserPresentable(this.helpMessage());
         }
 
+        const index = userIndex - 1;
```
Comment (TeakayAero) [inline src/commands/promote.ts]: ?

make index a non-const solves this

File: src/database/guild-datasource.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    private Playlist: Model<IPlaylist>;
+
+    constructor(guild: GuildManager) {
+        this.logger = getLogger(`GuildDataSource-${guild.id}`);
+        this.guild = guild;
+
+    }
+
+    async setupIfNecessary() {
+        if (!this.db) {
```
Comment (TeakayAero) [inline src/database/guild-datasource.ts]: log, error handle

Comment (TeakayAero) [reply src/database/guild-datasource.ts]: make this as `setup`, do this.db check in functions, prevent using unneeded await every time

File: src/database/guild-datasource.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        let songDoc: ISong;
+        while((songDoc = await songsCursor.next())) {
+            songs.push(songDoc.toSong(initiator));
+        }
+
+        this.logger.info(`Loaded songs from playlist ${listname}`);
+        return songs;
+    }
+
+    private async getPlaylist(name: string, user: User, create: boolean = false) {
```
Comment (TeakayAero) [inline src/database/guild-datasource.ts]: probably should specify return types

File: src/database/guild-datasource.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        const result = await this.Playlist.findOne({name});
+        if (result) {
+            return result;
+        } else if (create) {
+            return await new this.Playlist({
```
Comment (TeakayAero) [inline src/database/guild-datasource.ts]: no return await in async functions

File: src/database/mongo-service.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
+import {Logger, getLogger} from "../logger";
+import {connect, Mongoose} from "mongoose";
+
+class MongoDBService {
+    private uri: string = 'mongodb://localhost:27017';
```
Comment (TeakayAero) [inline src/database/mongo-service.ts]: config file

File: src/database/mongo-service.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            return false
+        }
+    }
+
+    async getConnection(database: string) {
+        const existConnection = this.client.connections.find((conn) => {
+            return conn.db.databaseName == database;
+        });
+        if (existConnection) return existConnection;
+        const connection = await this.client.createConnection(`${this.uri}/${database}`, {useNewUrlParser: true});
```
Comment (TeakayAero) [inline src/database/mongo-service.ts]: what does create Connection return on error?

File: src/database/schema.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
+import {Document, Schema} from "mongoose";
```
Comment (TeakayAero) [inline src/database/schema.ts]: Future: separate this file pls.

db/schema/ ...
db/service/ ....

**Review by TeakayAero — CHANGES_REQUESTED** [review]

File: src/app.ts (hunk 2 above)
Comment (JyaouShingan) [reply src/app.ts:14]: no handle, exit program

File: src/database/mongo-service.ts (hunk 13 above)
Comment (JyaouShingan) [reply src/database/mongo-service.ts]: throw

**Review by TeakayAero — APPROVED** [review]

**Pull request merged**